### ISSUE TITLE
Allow a single slot to be rendered for multiple times

### ISSUE THREAD

**Issue by Justineo** [vuejs/vue#8496]
### What problem does this feature solve?
Currently if we render a slot for multiple times (eg. inside a `v-for` loop), we'll get

```
Duplicate presence of slot "${name}" found in the same render tree - this will likely cause render errors.
```

I know it's intentional to prevent render errors (like losing correct data binding) but consider the following use case: I have a `v-breadcrumbs` component which has a default item separator of a text node `/`, and I want to allows users to specify their own separators which can be anything, like an `<v-icon>` component.

Currently if I want to make it work, I have to define a scoped slot in the `v-breadcrumbs` component and bind nothing to it:

```html
<slot name="sep" v-bind="{}">/</slot>
```

And component users must define `slot-scope` on it and not use anything from it:

```html
<v-breadcrumbs>
  <v-icon slot="sep" slot-scope="_" name="angle-right"/>
</v-breadcrumbs>
```

And currently Vue is using slots as fallbacks for scoped slots with the same name. Whether the users can use a slot doesn't rely on if they want data from slot scope, but on whether the slot is gonna be rendered for multiple times inside the component self, this may raise more confusion for our users (like #8175).

So why don't we just get rid of such caveats and let a slot to be rendered more than once? We can clone them on duplication to prevent render errors as we already run the scoped slot function each time anyway. This makes the logic a lot simpler for users IMO: if we do not expect to use data from the slot scope, just use a slot, otherwise we'll always have to declare `slot-scope` to create the binding.

In addition, this would make documenting the (scoped) slots of a component more consistent, we just declare the data structure for a slot scope and users can decide if they want some data to bind to the `slot-scope` and don't need to care about how many times it will be rendered.

### What does the proposed API look like?
No additional API needed, just clone the slot nodes upon render and remove the waring about duplicated slots.

It may be implemented in userland today in a quite hacky way, see: https://codesandbox.io/s/lp11y2wovz

I just tweaked the `cloneVNode` function from the Vue core a little bit and it seemed to be working as expected using render function (without losing reactivity and event binding). There maybe some other edge situations I missed so correct me if it's not such trivial.

**Comment by liximomo** [comment]
Why duplicate presence of slot would cause errors? I think it's common to resue slot. Maybe we should remove this warning.

**Comment by Justineo** [comment]
@liximomo See the link I posted earlier. It's not just easy as “removing the warning”.

**Comment by ky-is** [comment]
This is possible to do with render functions instead of templates. There may well be side effects to doing this, though it's worked for my use cases so far.

I do something similar in https://github.com/ky-is/vue-separator which coincidentally can be used for breadcrumb-style UI :)

**Comment by Justineo** [comment]
@ky-is

This is why Vue currently doesn't allow slots to be rendered multiple times as I understand. You are losing reactivity here.

See https://codesandbox.io/s/3kr81nl92p

**Comment by ky-is** [comment]
@Justineo Thanks for the link! Makes sense, I'll have to document that.

**Comment by Justineo** [comment]
@backbone87

It's actually very close to what I'm proposing (we may have to recursively clone all children as well).

As I understand this update isn't a breaking change. For library authors if they switch scoped slots to slots in some cases it also won't break those using templates. Library authors can also provide transparent updates for their users by manually fallback `$scopedSlots` to `$slots` if they are using render functions.

**Comment by backbone87** [comment]
I mean why have "normal" `$slots` anyway? From a dev history perspective `$slots` was just there before `$scopedSlots`, but in retrospect the scoped slot approach is just more flexible. But there is no need to have split approaches anymore. It is more work for the child component that has to check `$scopedSlots` *and* `$slot`. Just deprecate `$slots` altogether, mirror its properties into cloning functions of `$scopedSlots` and trigger deprecation when trying to use `$slots`. No library author needs to do anything in this case, except when making a release for the next major vue version, when the deprecated stuff gets kicked out.

**Comment by goodpixels** [comment]
Please, make this happen!

**Comment by backbone87** [comment]
another benefit of always using scopedSlots insteadof slots is lazy evaluation of slot content:
Imagine a component with the following template `loading(:while="!item") {{ item.name }}`
The loading comp wouldnt display the slot content until item becomes truthy, but "normal" slots are rendered eagerly causing errors despite the slot content not being displayed.

**Comment by Justineo** [comment]
Resolved by 530ca1b2db315fbd0e360807b2031d26665c5d3d and shipped with `vue@2.5.18-beta.0`. Cheers! 🎉
